### PR TITLE
Add rlottie

### PR DESCRIPTION
**Pull request by DreamOfIce**
Also see https://github.com/xmake-io/xmake-repo/issues/1830

Changes:
- Create package rlottie
- Add version `2019.02.07` for stb
- Add some sources and platform `wasm` for pximan
- Fix hash error of patches of openssl

### CODE REVIEW

**Comment by waruqi** [comment]
得提交到 dev 才能触发 ci tests

**Comment by SirLynix** [comment]
openssl patches checksum were good, it's probably git that changed your line endings on cloning (do you have core.autocrlf enabled?)

**Comment by DreamOfIce** [comment]
> 得提交到 dev 才能触发 ci tests

不小心忘记了 …

**Comment by DreamOfIce** [comment]
> openssl patches checksums were good, it's probably git that changed your line endings on cloning (do you have core.autocrlf enabled?)

At that time I tested under Linux and Windows did not match, let me check again

**Comment by DreamOfIce** [comment]
> > openssl patches checksums were good, it's probably git that changed your line endings on cloning (do you have core.autocrlf enabled?)
> 
> At that time I tested under Linux and Windows did not match, let me check again

After checking again and confirming that this is my problem, I have revert this

**Comment by DreamOfIce** [comment]
On the question of platforms supported by pixman, should we open a separate PR?

**Comment by xq114** [comment]
pixman archives from different source have difference sha256 sums. If the extra source urls are actually needed, please use alias to distinguish them and attach correct sha256 sums for each url

**Comment by DreamOfIce** [comment]
> pixman archives from different source have difference sha256 sums. If the extra source urls are actually needed, please use alias to distinguish them and attach correct sha256 sums for each url

Looking at it, gitlab and GitHub have the same sha256, but cairographics.org has a different value.
How about delete cairographics.org?

**Comment by xq114** [comment]
see https://github.com/xmake-io/xmake-repo/blob/master/packages/a/asio/xmake.lua

**Comment by DreamOfIce** [comment]
Also, why we test WebAssembly shared library building?
I remember that wasm does notsupport shared lib currently(still a draft)

**Comment by DreamOfIce** [comment]
Why meson in xmake does not support mingw?
https://github.com/xmake-io/xmake-repo/actions/runs/4215092585/jobs/7315986375#step:5:37

**Comment by waruqi** [comment]
> @waruqi Why meson in xmake does not support mingw? https://github.com/xmake-io/xmake-repo/actions/runs/4215092585/jobs/7315986375#step:5:37

Only the on_install of the meson package is not enabled for `@msys`, because I didn't have time to test it before, and there aren't too many packages that depend on meson to enable mingw.

**Comment by DreamOfIce** [comment]
> > @waruqi Why meson in xmake does not support mingw? https://github.com/xmake-io/xmake-repo/actions/runs/4215092585/jobs/7315986375#step:5:37
> 
> Only the on_install of the meson package is not enabled for `@msys`, because I didn't have time to test it before, and there aren't too many packages that depend on meson to enable mingw.

So I'll go enable it and try?

**Comment by DreamOfIce** [comment]
> > @waruqi Why meson in xmake does not support mingw? https://github.com/xmake-io/xmake-repo/actions/runs/4215092585/jobs/7315986375#step:5:37
> 
> Only the on_install of the meson package is not enabled for `@msys`, because I didn't have time to test it before, and there aren't too many packages that depend on meson to enable mingw.

Or for the time being directly with the rlottie comes with pixman?
After all that to test the full platform for meson

**Comment by waruqi** [comment]
> > > @waruqi Why meson in xmake does not support mingw? https://github.com/xmake-io/xmake-repo/actions/runs/4215092585/jobs/7315986375#step:5:37
> > 
> > 
> > Only the on_install of the meson package is not enabled for , because I didn't have time to test it before, and there aren't too many packages that depend on meson to enable mingw.`@msys`
> 
> So let me go enable it and try?

you can try, or you can open a new pr to improve meson first.

**Comment by DreamOfIce** [comment]
脚本域里面有什么办法检测设备是大端还是小端？
https://github.com/xmake-io/xmake-repo/actions/runs/4216703838/jobs/7319938325#step:5:278

**Comment by waruqi** [comment]
不能 只有编译器才知道，除非运行编译器去 dump 内置宏检测，没必要这么麻烦，应该在代码层面处理才对，可以打 patch

win arm 的搞不定，可以先不用管，目前还是实验性支持，先处理其他平台。

**Comment by DreamOfIce** [comment]
还有，这种是怎么回事？查过源文件确实包含了#include <limits>
https://github.com/xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375

**Comment by xq114** [comment]
> 脚本域里面有什么办法检测设备是大端还是小端？ [xmake-io/xmake-repo/actions/runs/4216703838/jobs/7319938325#step:5:278](https://github.com/xmake-io/xmake-repo/actions/runs/4216703838/jobs/7319938325#step:5:278)

可以认为xmake现在支持的平台都是小端，没有大端cpu

**Comment by DreamOfIce** [comment]
> > 脚本域里面有什么办法检测设备是大端还是小端？ [xmake-io/xmake-repo/actions/runs/4216703838/jobs/7319938325#step:5:278](https://github.com/xmake-io/xmake-repo/actions/runs/4216703838/jobs/7319938325#step:5:278)
> 
> 可以认为xmake现在支持的平台都是小端，没有大端cpu

不一定，arm架构同时支持大端和小端模式

**Comment by DreamOfIce** [comment]
> > 还有，Linux上这种是怎么回事？查过源文件确实包含了#include [xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375](https://github.com/xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375)
> 
> 可能是版本问题，master分支修复了

是不是因为workflow的缓存？

**Comment by waruqi** [comment]
> > > 还有，Linux上这种是怎么回事？查过源文件确实包含了#include [xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375](https://github.com/xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375)
> > 
> > 
> > 可能是版本问题，master分支修复了
> 
> 是不是因为workflow的缓存？

ci 上没配置缓存，你用的是 0.2 的 tag 。。不包含最新 master commit 

要么走 add_patches 对 0.2 打 patch ，要么切到 master commit

**Comment by DreamOfIce** [comment]
> > > > 还有，Linux上这种是怎么回事？查过源文件确实包含了#include [xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375](https://github.com/xmake-io/xmake-repo/actions/runs/4216703837/jobs/7319937927#step:6:375)
> > > 
> > > 
> > > 可能是版本问题，master分支修复了
> > 
> > 
> > 是不是因为workflow的缓存？
> 
> ci 上没配置缓存，你用的是 0.2 的 tag 。。不包含最新 master commit
> 
> 要么走 add_patches 对 0.2 打 patch ，要么切到 master commit

我还以为是说xmake的问题，在xmake的master分支修复了……
要不我加一个版本0.2.0-patch1？

**Comment by DreamOfIce** [comment]
目前cross编译遇到的[以下问题](https://github.com/xmake-io/xmake-repo/actions/runs/4220637114/jobs/7327105144#step:5:211) :
```
lottieanimation.cpp:(.text._ZNSt23_Sp_counted_ptr_inplaceI10RenderTaskSaIS0_ELN9__gnu_cxx12_Lock_policyE1EE10_M_disposeEv[_ZNSt23_Sp_counted_ptr_inplaceI10RenderTaskSaIS0_ELN9__gnu_cxx12_Lock_policyE1EE10_M_disposeEv]+0x368): undefined reference to `std::__atomic_futex_unsigned_base::_M_futex_notify_all(unsigned int*)'

```
参考 https://github.com/richfelker/musl-cross-make/issues/79 ，也许只能取消对arm-v7以下的支持？
如果要这样，怎么在on_install里面排除部分架构

**Comment by waruqi** [comment]
cross 的 arch 比较散，只能整个 disable

**Comment by DreamOfIce** [comment]
> cross 的 arch 比较散，只能整个 disable

能单独排除cross吗？

**Comment by waruqi** [comment]
> > cross 的 arch 比较散，只能整个 disable
> 
> 能单独排除cross吗？

on_install 展开，去掉 cross 就行了，参考其他包

**Comment by DreamOfIce** [comment]
> Also, why we test WebAssembly shared library building?
> I remember that wasm does notsupport shared lib currently(still a draft)

@waruqi 这个应该是ci的逻辑问题吧？wasm都不支持链接动态库

**Comment by waruqi** [comment]
> > Also, why we test WebAssembly shared library building?
> > I remember that wasm does notsupport shared lib currently(still a draft)
> 
> @waruqi 这个应该是ci的逻辑问题吧？wasm都不支持链接动态库

但是 shared config 是内置的，你避免不了用户不会去设置，以及一些静态包 add_deps 时候，内部已经开启了 shared，暂时先通过包层面禁用

https://github.com/xmake-io/xmake-repo/blob/bf29ad7459b81ca6847e2a4a6c5c867b558f1b36/packages/l/libsdl_gfx/xmake.lua#L28

**Comment by DreamOfIce** [comment]
> 不能 只有编译器才知道，除非运行编译器去 dump 内置宏检测，没必要这么麻烦，应该在代码层面处理才对，可以打 patch
> 
> win arm 的搞不定，可以先不用管，目前还是实验性支持，先处理其他平台。

这个应该是代码内部检测逻辑测不出来导致的，要不参考这个写一个：

https://github.com/ToxicFrog/vstruct/blob/b6889e3b275e9f9532b5dff4917e903b17e22730/io/endianness.lua#L35-L39

**Comment by DreamOfIce** [comment]
> > > Also, why we test WebAssembly shared library building?
> > > I remember that wasm does notsupport shared lib currently(still a draft)
> > 
> > 
> > @waruqi 这个应该是ci的逻辑问题吧？wasm都不支持链接动态库
> 
> 但是 shared config 是内置的，你避免不了用户不会去设置，以及一些静态包 add_deps 时候，内部已经开启了 shared，暂时先通过包层面禁用
> 
> https://github.com/xmake-io/xmake-repo/blob/bf29ad7459b81ca6847e2a4a6c5c867b558f1b36/packages/l/libsdl_gfx/xmake.lua#L28

没看懂……怎么个搞法？

**Comment by DreamOfIce** [comment]
还有就是这种复制文件的时候失败是怎么回事，权限问题？
https://github.com/xmake-io/xmake-repo/actions/runs/4220822433/jobs/7327459155#step:5:258

**Comment by waruqi** [comment]
> > > > Also, why we test WebAssembly shared library building?
> > > > I remember that wasm does notsupport shared lib currently(still a draft)
> > > 
> > > 
> > > @waruqi 这个应该是ci的逻辑问题吧？wasm都不支持链接动态库
> > 
> > 
> > 但是 shared config 是内置的，你避免不了用户不会去设置，以及一些静态包 add_deps 时候，内部已经开启了 shared，暂时先通过包层面禁用
> > https://github.com/xmake-io/xmake-repo/blob/bf29ad7459b81ca6847e2a4a6c5c867b558f1b36/packages/l/libsdl_gfx/xmake.lua#L28
> 
> 没看懂……怎么个搞法？

照搬贴过来不就好了

**Comment by waruqi** [comment]
> https://github.com/xmake-io/xmake-repo/actions/runs/4220822433/jobs/7327459155#step:5:258

to "/usr/lib/pkgconfig/rlottie.pc": Success.

装错地方了，得装到 package:installdir("lib") ，调下 cmakelists.txt 看下漏设置了什么，或者是否写死了，需要 patch

**Comment by DreamOfIce** [comment]
> > https://github.com/xmake-io/xmake-repo/actions/runs/4220822433/jobs/7327459155#step:5:258
> 
> to "/usr/lib/pkgconfig/rlottie.pc": Success.
> 
> 装错地方了，得装到 package:installdir("lib") ，调下 cmakelists.txt 看下漏设置了什么，或者是否写死了，需要 patch

应该是[这一](https://github.com/Samsung/rlottie/blob/master/CMakeLists.txt#L153):
```
INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.pc DESTINATION ${LIB_INSTALL_DIR}/pkgconfig)
```

File: packages/r/rlottie/xmake.lua [hunk 1]
```diff
@@ -24,13 +24,13 @@ package("rlottie")
 
     on_install("windows", "linux", "macosx", "android", "iphoneos", "watchos", "wasm", function (package)
         if package:arch("arm.*") then
-            package:add("defines","RAPIDJSON_ENDIAN=" .. string.byte(string.serialize(function() end), 7) == 0x00 and "RAPIDJSON_BIGENDIAN" or "RAPIDJSON_LITTLEENDI    AN    ")
+            package:add("defines","RAPIDJSON_ENDIAN=" .. (string.byte(string.serialize(function() end), 7) == 0x00 and "RAPIDJSON_BIGENDIAN" or "RAPIDJSON_LITTLEENDIAN"))
```
Comment (waruqi) [inline packages/r/rlottie/xmake.lua:27]: 这个不对吧，你检测 xmake 当前运行的 enbian 。。不代表 arm 对应的 target 一定也是这个 endian 。。交叉编译的 endian 为啥会去判断 host 的 endian?

Comment (DreamOfIce) [reply packages/r/rlottie/xmake.lua:27]: > 这个不对吧，你检测 xmake 当前运行的 enbian 是当前 host 的 。。不代表 arm 对应的 target 一定也是这个 endian 。。交叉编译的 endian 为啥会去判断 host 的 endian?

那还是尝试patch吧

Comment (DreamOfIce) [reply packages/r/rlottie/xmake.lua:27]: 相关的代码在这里
https://github.com/Tencent/rapidjson/blob/master/include/rapidjson/rapidjson.h#L227-L282

Comment (waruqi) [reply packages/r/rlottie/xmake.lua:27]: win arm 它不是写死 little 了么，你自己调下 我暂时没空看这个

https://github.com/Tencent/rapidjson/blob/012be8528783cdbf4b7a9e64f78bd8f056b97e24/include/rapidjson/rapidjson.h#L276

Comment (DreamOfIce) [reply packages/r/rlottie/xmake.lua:27]: 主要问题在于win arm上会报错检测不出来，因为win arm似乎都是小端：https://devblogs.microsoft.com/oldnewthing/20220726-00/?p=106898
我试下手动定义_LITTLE_ENDIAN

Comment (DreamOfIce) [reply packages/r/rlottie/xmake.lua]: 似乎没效果，是不是要加在rapidjson那边？虽然只是单纯地复制头文件

**Comment by DreamOfIce** [comment]
MacOS到底是什么错误，日志看不出来：
https://github.com/xmake-io/xmake-repo/actions/runs/4223042964/jobs/7332292287#step:4:289

**Comment by waruqi** [comment]
> MacOS到底是什么错误，日志看不出来： https://github.com/xmake-io/xmake-repo/actions/runs/4223042964/jobs/7332292287#step:4:289

这是 ios

> error: @programdir/core/sandbox/modules/os.lua:268: Undefined symbols for architecture arm64:
  "_pixman_composite_over_n_8888_asm_neon", referenced from:
      color_SourceOver(unsigned int*, int, unsigned int, unsigned int) in librlottie.a(vdrawhelper_neon.cpp.o)
  "_pixman_composite_src_n_8888_asm_neon", referenced from:
      memfill32(unsigned int*, unsigned int, int) in librlottie.a(vdrawhelper_neon.cpp.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by DreamOfIce** [comment]
这个应该在 https://github.com/Samsung/rlottie/blob/master/src/vector/pixman/pixman-arm-neon-asm.S 里面定义过了，还是因为 $ARCH 不对导致没包含进去？

File: packages/r/rlottie/xmake.lua (hunk 1 above)
Comment (DreamOfIce) [reply packages/r/rlottie/xmake.lua]: 解决了，测试通过

**Comment by DreamOfIce** [comment]
关于arm以及cross编译，实际上rlottie是支持的，但是要用meson：
https://github.com/Samsung/rlottie/blob/master/arm_build.sh

**Comment by waruqi** [comment]
先把 macos 等常用 pc 平台搞定， arm 什么的如果短期不行，可以先禁用掉相关平台

**Comment by DreamOfIce** [comment]
> 这个应该在 https://github.com/Samsung/rlottie/blob/master/src/vector/pixman/pixman-arm-neon-asm.S 里面定义过了，还是因为 $ARCH 不对导致没包含进去？

应该是编译平台和目标平台不一致的原因，rlottie按照编译的主机平台来决定是否添加这个文件

**Comment by DreamOfIce** [comment]
stb-image提供的.S文件似乎有问题，会报错
其中arm64-v8不用.S文件的时候构建成功，用了反而失败了。

**Comment by DreamOfIce** [comment]
这又是怎么回事？
```
ERROR: No host machine compiler for 'pixman/pixman-arma64-neon-asm.S'
```
https://github.com/xmake-io/xmake-repo/actions/runs/4231124359/jobs/7349253669#step:5:310

**Comment by DreamOfIce** [comment]
@waruqi 出什么事了嘛？Windows下载xmake全部404
https://github.com/xmake-io/xmake-repo/actions/runs/4232255113/jobs/7351762629

**Comment by waruqi** [comment]
> @waruqi 出什么事了嘛？Windows下载xmake全部404 https://github.com/xmake-io/xmake-repo/actions/runs/4232255113/jobs/7351762629

刚更新 master ，ci 上传跪了，重跑下

**Comment by DreamOfIce** [comment]
为什么手动指定cxx11后，macOS仍然无法识别列表初始化？
https://github.com/xmake-io/xmake-repo/actions/runs/4231124386/jobs/7349254523#step:4:392

**Comment by DreamOfIce** [comment]
> > @waruqi 出什么事了嘛？Windows下载xmake全部404 https://github.com/xmake-io/xmake-repo/actions/runs/4232255113/jobs/7351762629
> 
> 刚更新 master ，ci 上传跪了，重跑下

重跑我没权限，你来吧

**Comment by waruqi** [comment]
> > > @waruqi 出什么事了嘛？Windows下载xmake全部404 https://github.com/xmake-io/xmake-repo/actions/runs/4232255113/jobs/7351762629
> > 
> > 
> > 刚更新 master ，ci 上传跪了，重跑下
> 
> 重跑我没权限，你来吧

已经给你重跑过了，另外，重新 push ，或者 close + reopen pr ，你这也能触发重跑

File: packages/r/rlottie/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        add_configs("shared", {description = "Build shared library.", default = false, type = "boolean", readonly = true})
+        add_patches("0.2", path.join(os.scriptdir(), "patches", "0.2", "wasm.diff"), "46dfc912a6fee937f30f82906ef359ca92d1c9837946377e3ea9b40d933cadeb")
+    end
+
+    add_deps("cmake")
+    add_deps("freetype", {configs = {zlib = false}})
+    add_deps("pixman ~0.42.0", "rapidjson ~1.1.0", "stb 2019.02.07")
+
+    on_install("windows", "linux", "macosx", "android", "iphoneos", "watchos", "wasm", function (package)
+        package:set("languages", "cxx11")
```
Comment (waruqi) [inline packages/r/rlottie/xmake.lua:27]: 这里不要指定，删了

File: packages/r/rlottie/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        for name, enabled in pairs(package:configs()) do
+            if not package:extraconf("configs", name, "builtin") then
+                table.insert(configs, "-DLOTTIE_" .. name:upper() .. "=" .. (enabled and "ON" or "OFF"))
+            end
+        end
+        import("package.tools.cmake").install(package, configs)
+    end)
+
+    on_test(function (package)
+        assert(package:has_cxxtypes("rlottie::Surface", {includes = "rlottie.h"}))
```
Comment (waruqi) [inline packages/r/rlottie/xmake.lua:42]: 这里指定下。
另外，has_cxxtypes 改成 has_cxxsnippets ，带入更多 link 检测
```
        assert(package:has_cxxtypes("rlottie::Surface", {includes = "rlottie.h", configs = {languages = "c++11"}}))
```

**Comment by waruqi** [comment]
macos 给你过掉了，android 上，meson 目前这块还没适配，得改下 xmake 这里，适配下 meson for android 

https://github.com/xmake-io/xmake/blob/182613fd26ca880aa925fc02541cbf7f51104dd4/xmake/modules/package/tools/meson.lua#L173

**Comment by waruqi** [comment]
win static 库时候，这里不应该设置 dll import 

https://github.com/Samsung/rlottie/blob/056c0a1bc14bcee6f6b1a821013872a1228b0fa9/inc/rlottie.h#L34

感觉得处理下。

**Comment by DreamOfIce** [comment]
我加个patch？

**Comment by waruqi** [comment]
> 我加个patch？

直接走 io.replace 就地替换就行了（参考其他包），因为 shared 下正常不用 patch，仅仅对 static 时候才需要处理下

**Comment by DreamOfIce** [comment]
好了，顺路把wasm的patch也换成了io.replace

**Comment by DreamOfIce** [comment]
为什么会出现shared和static都为true？

**Comment by waruqi** [comment]
> 为什么会出现shared和static都为true？

?

**Comment by DreamOfIce** [comment]
https://github.com/xmake-io/xmake-repo/actions/runs/4240479170/jobs/7369542819#step:5:380

**Comment by DreamOfIce** [comment]
https://github.com/xmake-io/xmake-repo/actions/runs/4240479170/jobs/7369542819#step:5:380

**Comment by waruqi** [comment]
> https://github.com/xmake-io/xmake-repo/actions/runs/4240479170/jobs/7369542819#step:5:380

是这个库，切到 static 编译，但是里面似乎有个插件，还是强制走的 dll 编译。。rlottie-image-loader.dll

xmake 从 lib 目录下，同时探测到了 static/shared 。。所以都标记了

**Comment by waruqi** [comment]
https://github.com/Samsung/rlottie/blob/056c0a1bc14bcee6f6b1a821013872a1228b0fa9/CMakeLists.txt#L38

**Comment by DreamOfIce** [comment]
那怎么判断好？还是不管类型直接替换掉？

**Comment by waruqi** [comment]
> 那怎么判断好？还是不管类型直接替换掉？

还是原样 package:config("shared") 判断就行了，rlottie 切了，正常。。只是它附带了一个 dll 插件而已，不用管它。。

**Comment by DreamOfIce** [comment]
目前这样似乎替换不成功，要不我改成`package:config("static")`

**Comment by DreamOfIce** [comment]
似乎io.replace没效果，是哪里写错了吗？

File: packages/r/rlottie/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        table.insert(configs, "-DLIB_INSTALL_DIR=" .. package:installdir("lib"))
+        for name, enabled in pairs(package:configs()) do
+            if not package:extraconf("configs", name, "builtin") then
+                table.insert(configs, "-DLOTTIE_" .. name:upper() .. "=" .. (enabled and "ON" or "OFF"))
+            end
+        end
+        if package:is_plat("wasm") then
+            io.replace("CMakeLists.txt", "-Wl,--no-undefined", "-Wl")
+        elseif package:is_plat("windows") and package:config("static") then
+            io.replace("inc/rlottie.h", "#define RLOTTIE_API __declspec(dllimport)", "#define RLOTTIE_API")
```
Comment (waruqi) [inline packages/r/rlottie/xmake.lua:40]: 参考其他包，最后追加 {plain = true} 否则走的 lua 模式匹配

Comment (SirLynix) [reply packages/r/rlottie/xmake.lua:40]: 该测试应该是 `not package:config("shared")`。

Comment (DreamOfIce) [reply packages/r/rlottie/xmake.lua]: Done

**Comment by DreamOfIce** [comment]
@waruqi FreeBSD这个是怎么回事？重跑后还是不行

**Comment by waruqi** [comment]
那个 ci 不稳定 不用管它。

**Pull request merged**